datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -500,302 +387,344 @@ fn optimize_join(
-                    source: source.clone(),
-                    projection: projection.clone(),
-                    projected_schema: projected_schema.clone(),
-                    table_name: table_name.clone(),
-                    filters: new_filters,
-                    fetch: *fetch,
-                }),
-            )
-        }
-        _ => {
-            // all other plans are _not_ filter-commutable
-            let used_columns = plan
-                .schema()
-                .fields()
-                .iter()
-                .map(|f| f.qualified_column())
-                .collect::<HashSet<_>>();
-            issue_filters(state, used_columns, plan)
-        }
-    }
-}
+                push_down_all_join(predicates, filter.input(), left, right, vec![])?
+            }
+            LogicalPlan::TableScan(scan) => {
+                let mut new_scan_filters = scan.filters.clone();
+                let mut new_predicate = vec![];
+
+                let filter_predicates = utils::split_conjunction_owned(
+                    utils::cnf_rewrite(filter.predicate().clone()),
+                );
+
+                for filter_expr in &filter_predicates {
+                    let (preserve_filter_node, add_to_provider) =
+                        match scan.source.supports_filter_pushdown(filter_expr)? {
+                            TableProviderFilterPushDown::Unsupported => (true, false),
+                            TableProviderFilterPushDown::Inexact => (true, true),
+                            TableProviderFilterPushDown::Exact => (false, true),
+                        };
+                    if preserve_filter_node {
+                        new_predicate.push(filter_expr.clone());
+                    }
+                    if add_to_provider {
+                        // avoid reduplicated filter expr.
+                        if new_scan_filters.contains(filter_expr) {
+                            continue;
+                        }
+                        new_scan_filters.push(filter_expr.clone());
+                    }
+                }
 
-impl OptimizerRule for FilterPushDown {
-    fn name(&self) -> &str {
-        "filter_push_down"
-    }
+                let new_scan = LogicalPlan::TableScan(TableScan {
+                    source: scan.source.clone(),
+                    projection: scan.projection.clone(),
+                    projected_schema: scan.projected_schema.clone(),
+                    table_name: scan.table_name.clone(),
+                    filters: new_scan_filters,
+                    fetch: scan.fetch,
+                });
+
+                match conjunction(new_predicate) {
+                    Some(predicate) => LogicalPlan::Filter(Filter::try_new(
+                        predicate,
+                        Arc::new(new_scan),
+                    )?),
+                    None => new_scan,
+                }
+            }
+            _ => plan.clone(),
+        };
 
```

Comment:
I think the `LogicalPlan::Repartition `is also filter commutable. it is missing case.

Reply:
resolved